orderbook/src/database/presignatures.rs

```diff
@@ -0,0 +1,375 @@
+use crate::conversions::h160_from_vec;
+use crate::database::Postgres;
+use anyhow::{anyhow, Context, Result};
+use ethcontract::H160;
+use futures::{
+    future,
+    stream::{BoxStream, StreamExt, TryStreamExt},
+};
+use model::{order::OrderUid, presignature::PreSignature};
+use std::convert::TryInto;
+
+#[async_trait::async_trait]
+pub trait PreSignatureRetrieving: Send + Sync {
+    fn presignatures<'a>(
+        &'a self,
+        filter: &'a PreSignatureFilter,
+    ) -> BoxStream<'a, Result<PreSignature>>;
+}
+
+/// Any default value means that this field is unfiltered.
+#[derive(Debug, Default, PartialEq)]
+pub struct PreSignatureFilter {
+    pub owner: Option<H160>,
+    pub order_uid: Option<OrderUid>,
+}
+
+impl PreSignatureRetrieving for Postgres {
+    fn presignatures<'a>(
+        &'a self,
+        filter: &'a PreSignatureFilter,
+    ) -> BoxStream<'a, Result<PreSignature>> {
+        const QUERY: &str = "\
+            SELECT DISTINCT ON (t.order_uid) \
+                t.block_number, \
+                t.log_index, \
+                t.owner, \
+                t.order_uid, \
+                t.signed \
+            FROM presignatures t \
+            WHERE \
+                ($1 IS NULL OR t.owner = $1) \
+            AND \
+                ($2 IS NULL OR t.order_uid = $2) \
+            ORDER BY \
+                t.order_uid, t.block_number DESC, t.log_index DESC;";
```

Comment:
Is this needed? I don't think it is a problem just wondering why. Or just to make it deterministic? On that note maybe we should not allow getting *all* pre signatures without any filter in the http api handler just like for orders and trades because it would be very expensive for the database.

Reply:
This is important and goes together with `DISTINCT ON` at the beginning: given some uid `uid`, I only want to read the most recent entry that contains `uid`, as the latest presignature for `uid` invalidates all past ones. This explains `ORDER BY t.block_number DESC, t.log_index DESC`, the reason why there is also `t.order_uid` is that the query appears to be invalid if the `DISTINCT ON` parameter is not in the sort values. 

> we should not allow getting all pre signatures without any filter in the http api handler

This PR should already require the query to have a non-empty filter if called from the API. I left the option of retrieving everything from an internal call to this function available, should I change this?

Reply:
You're right. You shouldn't change it. Imo should only be checked in the api and not in the database trait, like it already is.